src/java/org/apache/cassandra/db/virtual/AbstractSchemaMetadataTable.java

```diff
@@ -0,0 +1,299 @@
+                           .addClusteringColumn(COLUMN_NAME, UTF8Type.instance)
+                           .addClusteringColumn(UDT_NAME, UTF8Type.instance)
+                           .addRegularColumn(schemaTableType.columnName, UTF8Type.instance)
+                           .build());
+        this.schemaTableType = schemaTableType;
+    }
+
+    /**
+     * Extract metadata from a keyspace.
+     * @param keyspace the keyspace metadata
+     * @return the metadata value, or null if not present
+     */
+    protected abstract String extractKeyspaceMetadata(KeyspaceMetadata keyspace);
+
+    /**
+     * Extract metadata from a table.
+     * @param table the table metadata
+     * @return the metadata value, or null if not present
+     */
+    protected abstract String extractTableMetadata(TableMetadata table);
+
+    /**
+     * Extract metadata from a column.
+     * @param column the column metadata
+     * @return the metadata value, or null if not present
+     */
+    protected abstract String extractColumnMetadata(ColumnMetadata column);
+
+    /**
+     * Extract metadata from a UDT.
+     * @param udt the user-defined type
+     * @return the metadata value, or null if not present
+     */
+    protected abstract String extractUdtMetadata(UserType udt);
+
+    @Override
+    public DataSet data()
+    {
+        SimpleDataSet result = new SimpleDataSet(metadata());
+        SchemaProvider schemaProvider = Schema.instance;
+        for (String keyspaceName : schemaProvider.getKeyspaces())
+        {
+            KeyspaceMetadata keyspace = schemaProvider.getKeyspaceMetadata(keyspaceName);
+            if(keyspace == null)
+            {
+                noSpamLogger.warn("Keyspace metadata not found for keyspace: {}", keyspaceName);
+                continue;
+            }
+            addKeyspaceRow(result, keyspace);
+
+            for (TableMetadata table : keyspace.tables)
+            {
+                addTableRow(result, keyspace, table);
+
+                for (ColumnMetadata column : table.columns())
+                {
+                    addColumnRow(result, keyspace, table, column);
+                }
+            }
+
+            for (UserType udt : keyspace.types)
+            {
+                addUdtRow(result, keyspace, udt);
+            }
+        }
+
+        return result;
+    }
+
+    @Override
+    public DataSet data(DecoratedKey partitionKey)
+    {
+        SimpleDataSet result = new SimpleDataSet(metadata());
+
+        ByteBuffer key = partitionKey.getKey();
+        ByteBuffer[] components = ((CompositeType) metadata().partitionKeyType).split(key);
+        String objectType = UTF8Type.instance.compose(components[0]);
+        String keyspaceName = UTF8Type.instance.compose(components[1]);
+
+        KeyspaceMetadata keyspace = Schema.instance.getKeyspaceMetadata(keyspaceName);
+        if (keyspace == null)
+        {
+            logger.debug("Keyspace not found: {}", keyspaceName);
```

Comment:
same ... not sure what is this good for

Reply:
Removed logs and instead of returning zero results, I am now failing the query with appropriate error message.